storm-core/src/clj/org/apache/storm/daemon/nimbus.clj

```diff
@@ -1008,23 +1008,24 @@
       (reset! (:id->worker-resources nimbus) {}))
     ;; tasks figure out what tasks to talk to by looking at topology at runtime
     ;; only log/set when there's been a change to the assignment
-    (doseq [[topology-id assignment] new-assignments
-            :let [existing-assignment (get existing-assignments topology-id)
-                  topology-details (.getById topologies topology-id)]]
-      (if (= existing-assignment assignment)
-        (log-debug "Assignment for " topology-id " hasn't changed")
-        (do
-          (log-message "Setting new assignment for topology id " topology-id ": " (pr-str assignment))
-          (.setAssignment storm-cluster-state topology-id (thriftify-assignment assignment))
-          )))
-    (->> new-assignments
-          (map (fn [[topology-id assignment]]
-            (let [existing-assignment (get existing-assignments topology-id)]
-              [topology-id (map to-worker-slot (newly-added-slots existing-assignment assignment))]
-              )))
-          (into {})
-          (.assignSlots inimbus topologies)))
-    (log-message "not a leader, skipping assignments")))
+    (locking (:sched-lock nimbus)
```

Comment:
Couldn't a wrong ordering of events happen  since we are locking when calculating a scheduling then unlocking and then locking and uploading the new scheduling and unlocking
for example:
T0: submit
T1: rebalance
T2: rebalance - calculate new scheduling
T3: submit - calculate new scheduling
T4: rebalance - upload new scheduling to zk
T5: submit - upload new scheduling to zk

even though we should end up with the scheduling calculated by the rebalance but we end up with scheduling calculated from the original submit.